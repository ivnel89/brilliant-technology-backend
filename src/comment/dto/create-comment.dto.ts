import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUUID, Length } from "class-validator";

export class CreateCommentDto {
  @ApiProperty()
  @IsString()
  @Length(1, 10000)
  content: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  authorId: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  articleId: string;

  @ApiProperty({
    nullable: true,
    required: false
  })
  @IsString()
  @IsUUID()
  @IsOptional()
  parentCommentId: string;
}
