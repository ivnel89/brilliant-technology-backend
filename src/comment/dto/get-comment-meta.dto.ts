import { ApiProperty } from "@nestjs/swagger";
import { IsArray, IsString, IsUUID } from "class-validator";

export class GetCommentsDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;

  @ApiProperty({
     isArray: true
  })
  @IsArray()
  commentIds: Array<string>;
}
