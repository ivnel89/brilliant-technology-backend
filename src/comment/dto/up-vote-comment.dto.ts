import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class UpVoteCommentDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;
}
