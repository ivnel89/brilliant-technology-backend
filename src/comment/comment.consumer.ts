import { InjectQueue, OnQueueActive, Process, Processor } from '@nestjs/bull';
import { Injectable } from '@nestjs/common';
import { Job, Queue } from 'bull';
import { QueueName } from 'src/config/queueName.enum';
import { CommentService } from './comment.service';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';

@Injectable()
@Processor(QueueName.COMMENTS_QUEUE)
export class CommentConsumer {
  constructor(
    private readonly commentService: CommentService,
  ) {}

  @OnQueueActive()
  onActive(job: Job) {
    console.log(
      `Processing job ${job.id} of type ${job.name} with data ${job.data}...`,
    );
  }

  @Process('addUpVote')
  async processAddUpVote(job: Job<{
    commentId: string,
    upVoteCommentDto: UpVoteCommentDto
  }>){
    const {
      data: { commentId, upVoteCommentDto },
    } = job;
    console.log('consuming job')
    this.commentService.addUpVote(commentId, upVoteCommentDto)
  }

  @Process('removeUpVote')
  async processRemoveUpVote(job: Job<{
    commentId: string,
    upVoteCommentDto: UpVoteCommentDto
  }>){
    const {
        data: { commentId, upVoteCommentDto },
      } = job;
      console.log('consuming job')
      this.commentService.removeUpVote(commentId, upVoteCommentDto)
  }
  
}
