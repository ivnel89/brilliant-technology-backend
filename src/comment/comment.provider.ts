import { InjectQueue } from '@nestjs/bull';
import { Injectable } from '@nestjs/common';
import { Queue } from 'bull';
import { QueueName } from 'src/config/queueName.enum';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';

@Injectable()
export class CommentProvider {
  constructor(
    @InjectQueue(QueueName.COMMENTS_QUEUE) private commentsQueue: Queue
  ) {}

  async addUpVoteJob(commentId: string, upVoteCommentDto: UpVoteCommentDto){
    console.log('adding job')
    await this.commentsQueue.add('addUpVote',{
      commentId,
      upVoteCommentDto
    })
  }

  async removeUpVoteJob(commentId: string, upVoteCommentDto: UpVoteCommentDto){
    console.log('adding job')
    await this.commentsQueue.add('removeUpVote', {
      commentId,
      upVoteCommentDto,
    });
  }
  
}
