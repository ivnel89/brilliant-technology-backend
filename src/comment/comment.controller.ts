import { Controller, Post, Body, Put, Param, Get, Query, } from '@nestjs/common';
import { CommentConsumer } from './comment.consumer';
import { CommentProvider } from './comment.provider';
import { CommentService } from './comment.service';
import { CommentContract } from './contract/comment.contract';
import { CreateCommentDto } from './dto/create-comment.dto';
import { GetCommentsDto } from './dto/get-comment-meta.dto';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';
import { Comment } from './entities/comment.entity';
import { CommentContractMapper } from './mapper/comment.mapper';

@Controller('comment')
export class CommentController {
  constructor(
    private readonly commentService: CommentService,
    private readonly commentProvider: CommentProvider
    ) {}
  private commentContractMapper = new CommentContractMapper();

  @Post()
  async create(
    @Body() createCommentDto: CreateCommentDto,
  ): Promise<CommentContract> {
    const comment = await this.commentService.create(createCommentDto);
    return this.commentContractMapper.build(comment);
  }

  @Put(':id/upvote/add')
  async addUpVote(
    @Param('id') id: string,
    @Body() upVoteCommentDto: UpVoteCommentDto,
  ): Promise<CommentContract> {
    await this.commentProvider.addUpVoteJob(id, upVoteCommentDto);
    const comment = await this.commentService.findOne(id);
    return this.commentContractMapper.build(comment, upVoteCommentDto.userId);
  }

  @Put(':id/upvote/remove')
  async removeUpVote(
    @Param('id') id: string,
    @Body() upVoteCommentDto: UpVoteCommentDto,
  ): Promise<CommentContract> {
   await this.commentProvider.removeUpVoteJob(id, upVoteCommentDto);
    const comment = await this.commentService.findOne(id);
    return this.commentContractMapper.build(comment, upVoteCommentDto.userId);
  }

  @Get('')
  async getComments(
    @Query() getCommentMetaDto: GetCommentsDto,
  ): Promise<Array<CommentContract>>{
    let comments: Array<Comment>;
    if(getCommentMetaDto.articleId){
      comments = await this.commentService.getCommentsByArticleId(
        getCommentMetaDto.articleId,
      );
    }else{
      comments = await this.commentService.getComments(
        getCommentMetaDto.commentIds,
      );
    }
    return this.commentContractMapper.buildArray(
      comments,
      getCommentMetaDto.userId,
    );
  }
}
