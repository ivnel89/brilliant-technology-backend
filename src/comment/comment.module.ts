import { Module } from '@nestjs/common';
import { CommentService } from './comment.service';
import { CommentController } from './comment.controller';
import { UserModule } from 'src/user/user.module';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Comment } from './entities/comment.entity';
import { ArticleModule } from 'src/article/article.module';
import { CommentProvider } from './comment.provider';
import { BullModule } from '@nestjs/bull';
import { QueueName } from 'src/config/queueName.enum';
import { CommentConsumer } from './comment.consumer';

@Module({
  imports: [
    TypeOrmModule.forFeature([Comment]),
    UserModule,
    ArticleModule,
    BullModule.registerQueue(
      {
        name: QueueName.COMMENTS_QUEUE,
      },
    ),
  ],
  controllers: [CommentController],
  providers: [CommentService, CommentProvider, CommentConsumer],
})
export class CommentModule {}
