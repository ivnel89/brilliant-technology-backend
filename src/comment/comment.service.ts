import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { ArticleService } from 'src/article/article.service';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { CreateCommentDto } from './dto/create-comment.dto';
import { Comment } from './entities/comment.entity';

@Injectable()
export class CommentService {
  constructor(
    @InjectRepository(Comment)
    private commentRepository: Repository<Comment>,
    private readonly userService: UserService,
    private readonly articleService: ArticleService,
  ) {}

  async create(createCommentDto: CreateCommentDto) {
    const [ 
      user, 
      article
     ] = await Promise.all([
       this.userService.findOne(createCommentDto?.authorId),
       this.articleService.findOne(createCommentDto?.articleId)
      ])
    return this.commentRepository.save(
      new Comment(user, article, createCommentDto?.content),
    );
  }

}
