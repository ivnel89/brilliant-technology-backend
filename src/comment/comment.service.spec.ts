import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { ArticleService } from 'src/article/article.service';
import { ArticleRepository, mockArticle } from 'src/article/test/article.repository.mock';
import { mockUser, UserRepository } from 'src/user/test/user.repository.mock';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { CommentService } from './comment.service';
import { Comment } from './entities/comment.entity';
import { createCommentDto, mockCommentId, upVoteCommentDto } from './test/comment.dto.mock';
import { CommentRepository, mockComment } from './test/comment.repository.mock';

describe('CommentService', () => {
  let service: CommentService;
  let repository: Repository<Comment>;
  let repositorySaveSpy, repositoryFindSpy, repositoryFindOneSpy, repositorySoftDeleteSpy;
  let userService: UserService;
  let articleService: ArticleService;
  let userServiceFindOneSpy, articleServiceFindOneSpy;
  
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        CommentService,
        CommentRepository,
        UserService,
        UserRepository,
        ArticleService,
        ArticleRepository
      ],
    }).compile();

    service = module.get<CommentService>(CommentService);
    repository = module.get<Repository<Comment>>(
      getRepositoryToken(Comment),
    );
    userService = module.get<UserService>(UserService);
    articleService = module.get<ArticleService>(ArticleService);
    repositorySaveSpy = jest.spyOn(repository,'save');
    repositoryFindSpy = jest.spyOn(repository,'find');
    repositoryFindOneSpy = jest.spyOn(repository,'findOne');
    repositorySoftDeleteSpy = jest.spyOn(repository,'softDelete');
    userServiceFindOneSpy = jest.spyOn(userService,'findOne');
    articleServiceFindOneSpy = jest.spyOn(articleService,'findOne');
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('create', () => {
    const saveMockError = 'saveMockError';
    beforeEach(() => {
      userServiceFindOneSpy.mockResolvedValueOnce(mockUser);
      articleServiceFindOneSpy.mockResolvedValueOnce(mockArticle);
    })

    it("should call repository's save method with right params", async () => {
      await expect(service.create(createCommentDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(userServiceFindOneSpy).toBeCalledWith(
            createCommentDto.authorId,
          );
          expect(articleServiceFindOneSpy).toBeCalledWith(
            createCommentDto.articleId,
          );
          expect(repositorySaveSpy).toBeCalledWith(
            new Comment(mockUser, mockArticle, createCommentDto.content),
          );
        });
    });
    it('should throw an error when save throws an error', async () => {
      repositorySaveSpy.mockRejectedValueOnce(saveMockError);
      await expect(service.create(createCommentDto)).rejects.toBe(saveMockError);
    });
  });

  describe('addUpVote', () => {
    it('add upvoter user to list of upvoters & save', async () => {
      await expect(service.addUpVote(mockCommentId, upVoteCommentDto)).resolves.toMatchSnapshot().then(() => {
        expect(repositorySaveSpy).toBeCalledWith(mockComment);
      })
    })
  })

  describe('removeUpVote', () => {
    it('removes upvoter user to list of upvoters & save', async () => {
      await expect(service.removeUpVote(mockCommentId, upVoteCommentDto)).resolves.toMatchSnapshot().then(() => {
        expect(repositorySaveSpy).toBeCalledWith(mockComment);
      })
    })
  })
});
