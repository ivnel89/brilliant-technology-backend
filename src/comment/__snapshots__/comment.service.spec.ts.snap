// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentService addUpVote add upvoter user to list of upvoters & save 1`] = `
Comment {
  "author": User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
  "content": "comment content",
  "parentArticle": Article {
    "author": User {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "isActive": true,
      "lastName": "Bar",
    },
    "comments": Array [
      Comment {
        "author": User {
          "displayPicture": "www.example.com/img.png",
          "firstName": "Foo",
          "isActive": true,
          "lastName": "Bar",
        },
        "content": "comment content",
        "parentArticle": [Circular],
      },
    ],
    "content": "article content",
  },
  "upVoters": Array [
    User {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "isActive": true,
      "lastName": "Bar",
    },
  ],
}
`;

exports[`CommentService create should call repository's save method with right params 1`] = `
Comment {
  "author": User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
  "content": "comment content",
  "parentArticle": Article {
    "author": User {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "isActive": true,
      "lastName": "Bar",
    },
    "comments": Array [
      Comment {
        "author": User {
          "displayPicture": "www.example.com/img.png",
          "firstName": "Foo",
          "isActive": true,
          "lastName": "Bar",
        },
        "content": "comment content",
        "parentArticle": [Circular],
      },
    ],
    "content": "article content",
  },
}
`;

exports[`CommentService removeUpVote removes upvoter user to list of upvoters & save 1`] = `
Comment {
  "author": User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
  "content": "comment content",
  "parentArticle": Article {
    "author": User {
      "displayPicture": "www.example.com/img.png",
      "firstName": "Foo",
      "isActive": true,
      "lastName": "Bar",
    },
    "comments": Array [
      Comment {
        "author": User {
          "displayPicture": "www.example.com/img.png",
          "firstName": "Foo",
          "isActive": true,
          "lastName": "Bar",
        },
        "content": "comment content",
        "parentArticle": [Circular],
      },
    ],
    "content": "article content",
  },
  "upVoters": Array [],
}
`;
